sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/AzureKeyVaultConfigurationProvider.cs

```diff
@@ -0,0 +1,170 @@
+                {
+                    continue;
+                }
+
+                var secretId = secret.Name;
+                if (oldLoadedSecrets != null &&
+                    oldLoadedSecrets.TryGetValue(secretId, out var existingSecret) &&
+                    existingSecret.IsUpToDate(secret.UpdatedOn))
+                {
+                    oldLoadedSecrets.Remove(secretId);
+                    newLoadedSecrets.Add(secretId, existingSecret);
+                }
+                else
+                {
+                    tasks.Add(_client.GetSecretAsync(secret.Name));
+                }
+            }
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            foreach (var task in tasks)
+            {
+                var secretBundle = task.Result;
+                newLoadedSecrets.Add(secretBundle.Value.Name, new LoadedSecret(_manager.GetKey(secretBundle), secretBundle.Value.Value, secretBundle.Value.Properties.UpdatedOn));
+            }
+
+            _loadedSecrets = newLoadedSecrets;
+
+            // Reload is needed if we are loading secrets that were not loaded before or
+            // secret that was loaded previously is not available anymore
+            if (tasks.Any() || oldLoadedSecrets?.Any() == true)
+            {
+                SetData(_loadedSecrets, fireToken: oldLoadedSecrets != null);
+            }
+
+            // schedule a polling task only if none exists and a valid delay is specified
+            if (_pollingTask == null && _reloadInterval != null)
+            {
+                _pollingTask = PollForSecretChangesAsync();
+            }
+        }
+
+        private void SetData(Dictionary<string, LoadedSecret> loadedSecrets, bool fireToken)
+        {
+            var data = new Dictionary<string, string>(loadedSecrets.Count, StringComparer.OrdinalIgnoreCase);
+            foreach (var secretItem in loadedSecrets)
+            {
+                data.Add(secretItem.Value.Key, secretItem.Value.Value);
+            }
+
+            Data = data;
+            if (fireToken)
+            {
+                OnReload();
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            _cancellationToken.Cancel();
+        }
+
+        private readonly struct LoadedSecret
+        {
+            public LoadedSecret(string key, string value, DateTimeOffset? updated)
+            {
+                Key = key;
+                Value = value;
+                Updated = updated;
+            }
+
+            public string Key { get; }
+            public string Value { get; }
+            public DateTimeOffset? Updated { get; }
+
+            public bool IsUpToDate(DateTimeOffset? updated)
+            {
+                if (updated.HasValue != Updated.HasValue)
+                {
+                    return false;
+                }
+
+                return updated.GetValueOrDefault() == Updated.GetValueOrDefault();
```

Comment:
I'm not understanding why this is simple equality. Care to clarify?

Reply:
Updated value can only increase so equality is enough here, if we don't have the exact time it's not up-to-date.